src/httpClient.ts

```diff
@@ -51,142 +47,58 @@ export class HttpClient {
-        currentRetryDelay *= 2;
+    if (IS_DEBUG) {
+      console.log("\n");
+      console.debug(`Fetching URL: ${url}`);
+      console.debug(`Method: ${finalInit.method || "GET"}`);
+      if (finalInit.body) {
+        console.debug(`Body: ${finalInit.body}`);
       }
+      console.debug(`Headers: ${JSON.stringify(finalInit.headers, null, 2)}`);
     }
-    throw lastError;
-  }
-
-  private nonTransientError(response: Response): boolean {
-    return [UNAUTHORIZED, FORBIDDEN, NOT_FOUND, BAD_REQUEST].includes(
-      response.status,
-    );
-  }
-
-  private delay(ms: number) {
-    return new Promise((resolve) => setTimeout(resolve, ms));
-  }
-
-  private shouldRetry(response: Response): boolean {
-    return (
-      response.status === TOO_MANY_REQUESTS ||
-      response.status >= INTERNAL_SERVER_ERROR
-    );
-  }
-
-  // Network error check
-  private isNetworkError(error: any): boolean {
-    return error instanceof TypeError;
-  }
-
-  // This could be redundant if we sent the correct error message from the server
-  private generateErrorMessage(response: Response, url: string) {
-    const status = response.status;
-    const statusText = response.statusText;
-
-    let errorMessage = `Request failed with status ${status} ${statusText} while accessing ${url}.`;
-
-    if (status === UNAUTHORIZED) {
-      errorMessage = `UNAUTHORIZED: Invalid API key ending in '${this.apiKey.slice(
-        -4,
-      )}'. Visit https://app.hamming.ai/settings to see valid API keys.`;
-    } else if (status === FORBIDDEN) {
-      errorMessage = `FORBIDDEN: You do not have permission to access ${url}.`;
-    } else if (status === NOT_FOUND) {
-      errorMessage = `NOT FOUND: The requested resource at ${url} could not be found.`;
-    } else if (status === TOO_MANY_REQUESTS) {
-      errorMessage = `TOO MANY REQUESTS: You are being rate limited. Please wait before making additional requests. If the issue persists, feel free to email us at founders@hamming.ai for help.`;
-    } else if (status >= 500) {
-      errorMessage = `SERVER ERROR: There was a problem with the server while accessing ${url}. If the issue persists, feel free to email us at founders@hamming.ai for help.`;
-    }
-    return errorMessage;
-  }
-
-  private async attemptFetch(
-    url: string,
-    init?: RequestInit,
-  ): Promise<Response> {
-    const headers = this.createHeaders(init?.headers);
-    const response = await fetch(url, { ...init, headers });
-    return response;
-  }
 
-  private createHeaders(existingHeaders?: HeadersInit): HeadersInit {
-    const headers: HeadersInit = {
-      ...existingHeaders,
-      authorization: `Bearer ${this.apiKey}`,
-    };
-    if (!existingHeaders || !("content-type" in existingHeaders)) {
-      headers["content-type"] = "application/json";
+    const MAX_ATTEMPTS = this.retries;
+    const resp = await this.fetchClient.fetchRetry(url, {
+      ...finalInit,
+      retryOn: function (attempt, error, response) {
```

Comment:
this also could be an array instead of a method, easier to read

Reply:
Ahh; we'd need to change the base implementation for this. If we specify an array, it no longer retries on network errors but only if a status matches